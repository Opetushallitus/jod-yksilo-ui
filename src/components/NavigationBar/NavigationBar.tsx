import { LanguageButton, UserButton } from '@/components';
import { useMediaQueries } from '@jod/design-system';

export interface NavigationBarLinkProps {
  className?: string;
  role?: string;
  title?: string;
  children: React.ReactNode;
}

export type NavigationBarLink = React.ComponentType<NavigationBarLinkProps>;

export interface NavigationBarProps {
  /** Navigation logo */
  logo: React.ReactNode;
  /** Place for menu opener button */
  menuComponent?: React.ReactNode;
  /** For language selection button **/
  onLanguageClick: () => void;
  /** Navigation avatar */
  user?: {
    name: string;
    component: NavigationBarLink;
  };
}

/**
 * This component is a navigation bar that displays a logo, and an avatar.
 */
export const NavigationBar = ({ logo, menuComponent, onLanguageClick, user }: NavigationBarProps) => {
  const { sm } = useMediaQueries();

  return (
    <div className="min-w-min shadow-border bg-white">
      <nav
        role="navigation"
        className="mx-auto flex h-11 items-center gap-4 px-5 sm:px-7 py-3 font-semibold xl:container"
      >
        {logo}
        <ul className="inline-flex items-center gap-3 sm:gap-5 ml-auto">
          {menuComponent && <li className="ml-3 sm:ml-0">{menuComponent}</li>}
          {sm && (
            <>
              <li className="ml-5">
                <LanguageButton onClick={onLanguageClick} />
              </li>
              <li>
                <UserButton user={user} />
              </li>
            </>
          )}
        </ul>
      </nav>
    </div>
  );
};
