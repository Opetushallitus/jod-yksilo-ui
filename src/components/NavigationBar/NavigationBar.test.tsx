import { useMediaQueries } from '@jod/design-system';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { NavigationBar, NavigationBarLinkProps } from './NavigationBar';

vi.mock('@jod/design-system', () => ({
  useMediaQueries: vi.fn(),
}));

vi.mock('react-router-dom', () => ({
  useLocation: () => ({
    pathname: 'callback-url',
  }),
}));

describe('NavigationBar', () => {
  vi.mocked(useMediaQueries).mockReturnValue({ sm: true, md: false, lg: false, xl: false });

  const logo = <div>logo</div>;

  const user = {
    name: 'Reetta Räppänä',
    component: ({ children, ...rootProps }: NavigationBarLinkProps) => (
      <a href="/profile" aria-label="Profile" {...rootProps}>
        {children}
      </a>
    ),
  };

  const onLanguageClick = vi.fn();

  it('renders only user', () => {
    const { container } = render(<NavigationBar logo={logo} user={user} onLanguageClick={onLanguageClick} />);

    // Assert snapshot
    expect(container.firstChild).toMatchSnapshot();

    // Assert user
    const userAvatar = screen.queryByTitle(user.name);
    expect(userAvatar).toBeInTheDocument();
  });

  it('renders no navigation items and no user', () => {
    const { container } = render(<NavigationBar logo={logo} onLanguageClick={onLanguageClick} />);

    // Assert snapshot
    expect(container.firstChild).toMatchSnapshot();

    // Assert user is not rendered
    const userAvatar = screen.queryByTitle(user.name);
    expect(userAvatar).toBeNull();
  });

  it('renders menu component', () => {
    const menuComponent = <div>Menu</div>;
    render(<NavigationBar logo={logo} menuComponent={menuComponent} onLanguageClick={onLanguageClick} />);
    const menuComponentElement = screen.getByText('Menu');
    expect(menuComponentElement).toBeInTheDocument();
  });
});
