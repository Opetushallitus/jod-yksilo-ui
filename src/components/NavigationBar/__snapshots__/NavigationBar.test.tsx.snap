// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigationBar > renders no navigation items and no user 1`] = `
<div
  class="min-w-min shadow-border bg-white"
>
  <nav
    class="mx-auto flex h-11 items-center gap-4 px-5 sm:px-7 py-3 font-semibold xl:container"
    role="navigation"
  >
    <div>
      logo
    </div>
    <ul
      class="inline-flex items-center gap-3 sm:gap-5 ml-auto"
    >
      <li
        class="ml-5"
      >
        <button
          aria-label="Valitse kieli"
          class="flex h-8 w-8 items-center justify-center rounded-full bg-bg-gray-2"
        >
          <svg
            fill="currentColor"
            height="24"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm6.93 6h-2.95a15.65 15.65 0 0 0-1.38-3.56A8.03 8.03 0 0 1 18.92 8zM12 4.04c.83 1.2 1.48 2.53 1.91 3.96h-3.82c.43-1.43 1.08-2.76 1.91-3.96zM4.26 14C4.1 13.36 4 12.69 4 12s.1-1.36.26-2h3.38c-.08.66-.14 1.32-.14 2 0 .68.06 1.34.14 2H4.26zm.82 2h2.95c.32 1.25.78 2.45 1.38 3.56A7.987 7.987 0 0 1 5.08 16zm2.95-8H5.08a7.987 7.987 0 0 1 4.33-3.56A15.65 15.65 0 0 0 8.03 8zM12 19.96c-.83-1.2-1.48-2.53-1.91-3.96h3.82c-.43 1.43-1.08 2.76-1.91 3.96zM14.34 14H9.66c-.09-.66-.16-1.32-.16-2 0-.68.07-1.35.16-2h4.68c.09.65.16 1.32.16 2 0 .68-.07 1.34-.16 2zm.25 5.56c.6-1.11 1.06-2.31 1.38-3.56h2.95a8.03 8.03 0 0 1-4.33 3.56zM16.36 14c.08-.66.14-1.32.14-2 0-.68-.06-1.34-.14-2h3.38c.16.64.26 1.31.26 2s-.1 1.36-.26 2h-3.38z"
            />
          </svg>
        </button>
      </li>
      <li>
        <a
          aria-label="Kirjaudu sisään"
          class="flex h-8 w-8 items-center justify-center rounded-full bg-bg-gray-2"
          href="/login?lang=fi&callback=callback-url"
        >
          <svg
            fill="currentColor"
            height="24"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M12 6c1.1 0 2 .9 2 2s-.9 2-2 2-2-.9-2-2 .9-2 2-2m0 10c2.7 0 5.8 1.29 6 2H6c.23-.72 3.31-2 6-2m0-12C9.79 4 8 5.79 8 8s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm0 10c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`NavigationBar > renders only user 1`] = `
<div
  class="min-w-min shadow-border bg-white"
>
  <nav
    class="mx-auto flex h-11 items-center gap-4 px-5 sm:px-7 py-3 font-semibold xl:container"
    role="navigation"
  >
    <div>
      logo
    </div>
    <ul
      class="inline-flex items-center gap-3 sm:gap-5 ml-auto"
    >
      <li
        class="ml-5"
      >
        <button
          aria-label="Valitse kieli"
          class="flex h-8 w-8 items-center justify-center rounded-full bg-bg-gray-2"
        >
          <svg
            fill="currentColor"
            height="24"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm6.93 6h-2.95a15.65 15.65 0 0 0-1.38-3.56A8.03 8.03 0 0 1 18.92 8zM12 4.04c.83 1.2 1.48 2.53 1.91 3.96h-3.82c.43-1.43 1.08-2.76 1.91-3.96zM4.26 14C4.1 13.36 4 12.69 4 12s.1-1.36.26-2h3.38c-.08.66-.14 1.32-.14 2 0 .68.06 1.34.14 2H4.26zm.82 2h2.95c.32 1.25.78 2.45 1.38 3.56A7.987 7.987 0 0 1 5.08 16zm2.95-8H5.08a7.987 7.987 0 0 1 4.33-3.56A15.65 15.65 0 0 0 8.03 8zM12 19.96c-.83-1.2-1.48-2.53-1.91-3.96h3.82c-.43 1.43-1.08 2.76-1.91 3.96zM14.34 14H9.66c-.09-.66-.16-1.32-.16-2 0-.68.07-1.35.16-2h4.68c.09.65.16 1.32.16 2 0 .68-.07 1.34-.16 2zm.25 5.56c.6-1.11 1.06-2.31 1.38-3.56h2.95a8.03 8.03 0 0 1-4.33 3.56zM16.36 14c.08-.66.14-1.32.14-2 0-.68-.06-1.34-.14-2h3.38c.16.64.26 1.31.26 2s-.1 1.36-.26 2h-3.38z"
            />
          </svg>
        </button>
      </li>
      <li>
        <a
          aria-label="Profile"
          class="flex h-8 w-8 items-center justify-center rounded-full bg-secondary-3 text-black select-none"
          href="/profile"
          role="img"
          title="Reetta Räppänä"
        >
          RR
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
